Mathlib/CategoryTheory/Bicategory/Kan/HasKan.lean

```diff
@@ -0,0 +1,335 @@
+      g
+```
+-/
+scoped infixr:90 "⁺ " => lan
+
+@[simp]
+theorem lanLeftExtension_extension (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] :
+    (lanLeftExtension f g).extension = f⁺ g := rfl
+
+/-- The unit for the left Kan extension `f⁺ g`. -/
+def lanUnit (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] : g ⟶ f ≫ f⁺ g :=
+  (lanLeftExtension f g).unit
+
+@[simp]
+theorem lanLeftExtension_unit (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] :
+    (lanLeftExtension f g).unit = lanUnit f g := rfl
+
+/-- Evidence that the `Lan.extension f g` is a Kan extension. -/
+def lanIsKan (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] : (lanLeftExtension f g).IsKan :=
+  initialIsInitial
+
+variable {f : a ⟶ b} {g : a ⟶ c}
+
+/-- The family of 2-morphisms out of the left Kan extension `f⁺ g`. -/
+def lanDesc [HasLeftKanExtension f g] (s : LeftExtension f g) :
+    f⁺ g ⟶ s.extension :=
+  (lanIsKan f g).desc s
+
+@[reassoc (attr := simp)]
+theorem lanUnit_desc [HasLeftKanExtension f g] (s : LeftExtension f g) :
+    lanUnit f g ≫ f ◁ lanDesc s = s.unit :=
+  (lanIsKan f g).fac s
+
+@[simp]
+theorem lanIsKan_desc [HasLeftKanExtension f g] (s : LeftExtension f g) :
+    (lanIsKan f g).desc s = lanDesc s :=
+  rfl
+
+theorem Lan.existsUnique [HasLeftKanExtension f g] (s : LeftExtension f g) :
+    ∃! τ, lanUnit f g ≫ f ◁ τ = s.unit :=
+  (lanIsKan f g).existsUnique _
+
+/-- We say that a 1-morphism `h` commutes with the left Kan extension `f⁺ g` if the whiskered
+left extension for `f⁺ g` by `h` is a Kan extension of `g ≫ h` along `f`. -/
+class Lan.CommuteWith
+    (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g] {x : B} (h : c ⟶ x) : Prop where
+  commute : Nonempty <| IsKan <| (lanLeftExtension f g).whisker h
+
+namespace Lan.CommuteWith
+
+theorem of_isKan_whisker [HasLeftKanExtension f g] (t : LeftExtension f g) {x : B} (h : c ⟶ x)
+    (H : IsKan (t.whisker h)) (i : t.whisker h ≅ (lanLeftExtension f g).whisker h) :
+    Lan.CommuteWith f g h :=
+  ⟨⟨IsKan.ofIsoKan H i⟩⟩
+
+theorem of_lan_comp_iso [HasLeftKanExtension f g]
+    {x : B} {h : c ⟶ x} [HasLeftKanExtension f (g ≫ h)]
+    (i : f⁺ (g ≫ h) ≅ f⁺ g ≫ h)
+    (w : lanUnit f (g ≫ h) ≫ f ◁ i.hom = lanUnit f g ▷ h ≫ (α_ _ _ _).hom) :
+    Lan.CommuteWith f g h :=
+  ⟨⟨(lanIsKan f (g ≫ h)).ofIsoKan <| StructuredArrow.isoMk i⟩⟩
+
+variable (f : a ⟶ b) (g : a ⟶ c) [HasLeftKanExtension f g]
+variable {x : B} (h : c ⟶ x) [Lan.CommuteWith f g h]
+
+/-- Evidence that `h` commutes with the left Kan extension `f⁺ g`. -/
+def isKan : IsKan <| (lanLeftExtension f g).whisker h := Classical.choice Lan.CommuteWith.commute
+
+instance : HasLeftKanExtension f (g ≫ h) := (Lan.CommuteWith.isKan f g h).hasLeftKanExtension
+
+/-- If `h` commutes with `f⁺ g` and a left extension `t` is isomorphic to the left Kan extension
+of `g` along `f`, then `t.whisker h` is a left Kan extension. -/
+def isKanWhisker (t : LeftExtension f g) {x : B} (h : c ⟶ x) [Lan.CommuteWith f g h]
+    (i : lanLeftExtension f g ≅ t) :
```

Comment:
Here, I think it would be better if instead of an iso, the assumption was `t.IsUniversal`.